### PR TITLE
JP-3570: Match NaNs and DO_NOT_USE flags

### LINKED ISSUE #8345: Add SCI/ERR/DQ check to resample/cube_build
_Issue [JP-3570](https://jira.stsci.edu/browse/JP-3570) was created on JIRA by [David Law](https://jira.stsci.edu/secure/ViewProfile.jspa?name=dlaw):_

As seen in tickets such as <https://jira.stsci.edu/browse/JP-3250],> unexpected NaNs in the ERR or VAR arrays can cause problems downstream since those NaNs can be propagated to the resampled products.

This ticket is to therefore catch such issues earlier, by adding a check to the start of resample and cube_build to ensure that all input data with SCI=NaN also have ERR=NaN and DQ DO_NOT_USE bit set.  Likewise, ERR=NaN should have SCI=NaN and DQ DO_NOT_USE bit set.  I believe cube_build may already have something like this implemented, as it is fairly robust against NaN-valued inputs.  An equivalent may also be needed elsewhere in TSO3 processing since that does not use the resample step.

Once implemented, this will need some fairly thorough instrument-mode testing prior to being merged as it may reveal issues with reference files that need to be addressed first.

### PR DESCRIPTION
**Pull request by melanieclarke**
Resolves [JP-3570](https://jira.stsci.edu/browse/JP-3570)

Closes #8345 

Make a function that updates a datamodel with NaN values in data, error, or variance arrays whenever the DO_NOT_USE flag is set, and updates the dq array to DO_NOT_USE whenever a NaN value is found in data, error, or variance arrays.

Call the function on output data at the end of several key steps where DO_NOT_USE might be set: 
- flat_field
- pathloss
- photom
- outlier_detection

Also call it on input data at the beginning of a couple key steps where data are combined:
- resample/resample_spec
- cube_build

**Checklist for PR authors (skip items if you don't have permissions or they are not applicable)**
- [x] added entry in `CHANGES.rst` within the relevant release section
- [x] updated or added relevant tests
- [ ] ~updated relevant documentation~
- [x] added relevant milestone
- [x] added relevant label(s)
- [x] ran regression tests, post a link to the Jenkins job below.
      [How to run regression tests on a PR](https://github.com/spacetelescope/jwst/wiki/Running-Regression-Tests-Against-PR-Branches)
- [x] All comments are resolved
- [ ] Make sure the JIRA ticket is [resolved properly](https://github.com/spacetelescope/jwst/wiki/How-to-resolve-JIRA-issues)

### CODE REVIEW

**Comment by melanieclarke** [comment]
Started regression tests here:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1543/

I predict many new NaNs!

**Comment by melanieclarke** [comment]
Now that we are past build 11.0, I think this is ready for discussion and testing.

Regression test results:
- Many new NaNs in intermediate products in SCI/ERR/VAR arrays, most notably outputs from outlier_detection, flat_field, and photom.
- Imaging source catalogs show differences because NaNs are not ignored in photometry (e.g. [test_miri_image3_catalog](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1552/testReport/junit/jwst.regtest/test_miri_image/_stable_deps__test_miri_image3_catalog/))
- NIRSpec IFU shows changes in the DQ array after flat_field: NON_SCIENCE areas are now also flagged DO_NOT_USE.
- There are some small changes to values in the SURF_BRIGHT column only in x1d products (e.g. [test_nirspec_mos_spec3[v000000053-x1d]](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1552/testReport/junit/jwst.regtest/test_nirspec_mos_spec3/_stable_deps__test_nirspec_mos_spec3_v000000053_x1d_/))
- There are some small changes to weight values in s2d files (e.g. [test_nirspec_mos_spec3[v000000053-s2d]](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1552/testReport/junit/jwst.regtest/test_nirspec_mos_spec3/_stable_deps__test_nirspec_mos_spec3_v000000053_s2d_/))
- There are some changes to values in NIRCam coronography final products (e.g. [test_nircam_coron3_product[i2d]](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1552/testReport/junit/jwst.regtest/test_nircam_coron3/_stable_deps__test_nircam_coron3_product_i2d_/)).

File: jwst/lib/pipe_utils.py
Comment (emolter) [inline jwst/lib/pipe_utils.py:118]: Are there any cases where we would expect one of these arrays (say, `var_rnoise`) to be the wrong shape but still trust the results of this?  Put another way, isn't any shape mismatch a sign of a different issue earlier in the pipeline flow?  I guess I would expect this to throw an error.

File: jwst/lib/pipe_utils.py
Comment (emolter) [inline jwst/lib/pipe_utils.py:126]: would `is_invalid` ever be None?  The way I'm reading this, the array gets set on line 114 unless somehow none of the other extensions were found, correct?  And in that case the function seemingly has nothing to do since there are no extensions to match with the dq extension.

File: jwst/lib/pipe_utils.py
Comment (emolter) [inline jwst/lib/pipe_utils.py:136]: In what case would you want to pass a DataModel through here that has none of the extensions listed in line 108?

**Review by emolter — COMMENTED** [review]
Thanks Melanie this looks great already, just a two general questions:

- is the idea that the NaNs in the data and variance arrays are _always_ synced with the DO_NOT_USE flags in the DQ array?  If so, should we consider making some kind of check for this generic to all pipeline steps?  I guess it isn't obvious to me that the steps where you added this are the major/only places this would be helpful.
- I'm glad you checked edge cases where certain extensions are not found - not sure I would have thought of that - but why is the preferred behavior in those cases to return a copy of the datamodel instead of failing?

File: jwst/lib/pipe_utils.py
Comment (melanieclarke) [reply jwst/lib/pipe_utils.py:118]: This is not expected in normal use cases in the pipeline, but there are a couple unit tests for the resampling step that pass in a mismatched variance (with size 0) and expect the step to finish successfully, with a warning for the missing variance.

File: jwst/lib/pipe_utils.py
Comment (melanieclarke) [reply jwst/lib/pipe_utils.py:126]: That's true - if there are no extensions other than DQ, there's nothing to do.  I'll move the check for is_invalid = None higher up in the function.

File: jwst/lib/pipe_utils.py
Comment (melanieclarke) [reply jwst/lib/pipe_utils.py:136]: You wouldn't, really.  I just wanted it to make sure it was as general as possible, so it could be called everywhere.

**Comment by melanieclarke** [comment]
Thanks for reviewing @emolter.

> is the idea that the NaNs in the data and variance arrays are _always_ synced with the DO_NOT_USE flags in the DQ array?  If so, should we consider making some kind of check for this generic to all pipeline steps?  I guess it isn't obvious to me that the steps where you added this are the major/only places this would be helpful.

I think that's the idea, but I was reluctant to add it to every single pipeline step individually in the first pass.  I added it to steps that I know either modify DO_NOT_USE or set NaNs in data arrays.  I can add it to more if needed.

I didn't consider whether it could or should be done more generically, in a core function or class.  It does need some knowledge of the output model composition, so that it does the updates at the right level.  Do you have a better sense of where that might go?

> * I'm glad you checked edge cases where certain extensions are not found - not sure I would have thought of that - but why is the preferred behavior in those cases to return a copy of the datamodel instead of failing?

I was thinking that it should be as minimally invasive as possible, so it can be called on all kinds of datamodels, without assuming that they have all the extensions at all times.  As is, it could be used on partial copies of datamodels as well as on the final products that should have all the expected extensions.

File: jwst/lib/pipe_utils.py
Comment (melanieclarke) [reply jwst/lib/pipe_utils.py]: Done.

**Comment by emolter** [comment]
> I didn't consider whether it could or should be done more generically, in a core function or class. It does need some knowledge of the output model composition, so that it does the updates at the right level. Do you have a better sense of where that might go?

I don't know if I have the expertise or experience to answer the _should_ question, but I am sure it _could_ be implemented. 
 If that were the case I'd suggest putting it somewhere into `jwst.stpipe`, maybe `jwst.stpipe.utils`, and then calling it from the core JWST Step class.  It reminds me a bit of the `query_step_status` function we just implemented, and with regards to knowledge of the model composition, is it anything more complicated than what is implemented in that `query_step_status` function?  That is, do something to the model if it's not a Sequence, loop over all models in the sequence if it is?

File: jwst/outlier_detection/outlier_detection.py
Comment (braingram) [inline jwst/outlier_detection/outlier_detection.py:411]: To check that I'm following the changes. With this change calling outlier detection will now also assign `nan` to all pixels marked as outliers (whereas before it only updated the dq flags for those pixels)?

Comment (melanieclarke) [reply jwst/outlier_detection/outlier_detection.py:411]: Yes, that's correct.

Comment (braingram) [reply jwst/outlier_detection/outlier_detection.py:411]: Thanks! I didn't quite get that from the ticket which mentions setting err and dq when sci is nan or sci and dq when err is nan (but not setting sci and err when dq is do not use).

File: jwst/outlier_detection/outlier_detection_tso.py
Comment (braingram) [inline jwst/outlier_detection/outlier_detection_tso.py:74]: `self.input_models` is a `CubeModel` here. The iteration appears to produce the result of "keys" (so a `input_model` will be a string). Looking at `match_nans_and_flags` this should produce a warning?

Comment (melanieclarke) [reply jwst/outlier_detection/outlier_detection_tso.py:74]: Good catch, thanks!  In this case, it didn't throw a warning - it just quietly did nothing because it didn't have the attributes it was looking for.  Fixing this case now.

Comment (melanieclarke) [reply jwst/outlier_detection/outlier_detection_tso.py]: This should be fixed now.

**Comment by melanieclarke** [comment]
Per @drlaw1558 - we should keep the implementation limited to these specific steps for now, to avoid scope creep and possible unintended consequences.

**Review by emolter — APPROVED** [review]
all my questions are answered, thanks Melanie

File: jwst/lib/pipe_utils.py
Comment (braingram) [inline jwst/lib/pipe_utils.py:106]: Given the silent issue with the tso data for outlier detection. Would you add a `isinstance` check here to make sure the input is a `DataModel`?

Comment (melanieclarke) [reply jwst/lib/pipe_utils.py:106]: Sure.  I'll make that condition throw an error, since that's likely a code issue.

Comment (melanieclarke) [reply jwst/lib/pipe_utils.py:112]: Done, but I also had to allow datamodel ObjectNode instances as well - slits in MultSlitModels do not retain the SlitModel data type when they are appended to the `slits` list node.

Comment (braingram) [reply jwst/lib/pipe_utils.py:112]: Thanks!

**Review by drlaw1558 — APPROVED** [review]
LG2M.

**Comment by melanieclarke** [comment]
Thanks, @drlaw1558 for coordinating the testing!  I'll rebase and fix up the conflicts with later development and see if we can get this in.

**Comment by melanieclarke** [comment]
Outlier detection changed significantly after I initially wrote this PR, so I'm re-requesting review from @emolter and @braingram, if you have a chance.

I'll re-run regression tests shortly.

**Comment by melanieclarke** [comment]
Regression test results are inconclusive.  There are many new failures introduced by this branch, as expected, but it looks like there are also some unrelated failures that need okifying.  I will run again when things are more stable.

**Comment by melanieclarke** [comment]
I think the regression test output is as expected, compared to the previous run.

One thing I noticed in addition to the differences noted above: i2d and crf output for image3 is sometimes different (e.g. [test_miri_image3_i2d](https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1552/testReport/junit/jwst.regtest/test_miri_image/_stable_deps__test_miri_image3_i2d/)) because tweakreg is seeing a different set of sources to cross match on.  It might be worth looking into how the starfinder is handling NaNs: I wouldn’t have expected to see much difference here, since the DO_NOT_USE mask is already passed.

**Review by emolter — COMMENTED** [review]
It looks like `match_nans_and_flags` will not get called during the outlier detection step if `resample_data` is `False` for imaging or spectroscopic modes.  Is that true, and if so, is that the intended behavior?

**Comment by melanieclarke** [comment]
> It looks like `match_nans_and_flags` will not get called during the outlier detection step if `resample_data` is `False` for imaging or spectroscopic modes. Is that true, and if so, is that the intended behavior?

It should still be called in that case, via `flag_model_crs`.  I'll test.

**Comment by melanieclarke** [comment]
@emolter - confirmed, it's definitely called for image and spec outlier detection when resample_data is False.

@drlaw1558 - the new ref files are definitely confusing things in comparing to the current truth files, but I think there is additionally a change in tweakreg due to this branch. I ran the regtest locally with the same CRDS context on the current main and PR branches and compared the log output.

On main, tweakreg reports:
```
Detected 171 sources in jw01024001001_04101_00002_mirimage_cal.fits.
Detected 175 sources in jw01024001001_04101_00003_mirimage_cal.fits.
Detected 167 sources in jw01024001001_04101_00004_mirimage_cal.fits.
Detected 150 sources in jw01024001001_04101_00001_mirimage_cal.fits.
...
Final solution based on 111 objects.
```

On this branch, tweakreg reports:
```
Detected 155 sources in jw01024001001_04101_00002_mirimage_cal.fits.
Detected 161 sources in jw01024001001_04101_00003_mirimage_cal.fits.
Detected 153 sources in jw01024001001_04101_00004_mirimage_cal.fits.
Detected 149 sources in jw01024001001_04101_00001_mirimage_cal.fits.
...
Final solution based on 109 objects.
```

It's not a big change, and I don't think it should hold up this PR, but we might want to follow up on it later to make sure the star finder isn't handling NaNs badly.

File: jwst/resample/resample_step.py
Comment (emolter) [inline jwst/resample/resample_step.py:96]:
```suggestion
                input_models.shelve(model)
            del model
```

File: jwst/resample/resample_step.py
Comment (emolter) [inline jwst/resample/resample_step.py:93]: @melanieclarke I think these are the lines you were talking about in the stand-up.  The syntax looks correct.  If you wanted to, you could add a delete of `model` at the end, outside the `for` loop like I show in the suggestion, but I doubt it would matter much.

I think this is in an ok location too.  Ideally we would minimize the number of times models are iterated over because each of these is potentially a load and save operation when `on_disk` is True, but in this case I don't see a good place for this code inside a pre-existing loop in `ResampleData`.

That was a long way of saying LGTM! :)

File: jwst/resample/resample_step.py
Comment (melanieclarke) [reply jwst/resample/resample_step.py]: Done

File: jwst/resample/resample_step.py
Comment (melanieclarke) [reply jwst/resample/resample_step.py:93]: Thank you for reviewing again, hopefully for the last time!

**Comment by melanieclarke** [comment]
Regression tests one more time, to test the changes with the ModelLibrary work:
https://plwishmaster.stsci.edu:8081/job/RT/job/JWST-Developers-Pull-Requests/1689/

**Comment by melanieclarke** [comment]
Cross matching regression tests to the previous run, I don't see anything new or unexpected, so I think this is ready to go.

@tapastro, over to you!

**Review by tapastro — APPROVED** [review]
LGTM - I'll leave it to you to okify, if you have the time to do so!

**Comment by melanieclarke** [comment]
Okifying is done.  Let me know if you see problems.

**Pull request merged**
